### ISSUE TITLE
SAC policy problems?

### ISSUE THREAD

**Issue by albheim** [JuliaReinforcementLearning/ReinforcementLearning.jl#410]
I assume this refers to the dimensionality of the action? Using `dimension` in the docstring seems (to me) like you are asking for one specific dimension instead of the number of dimensions and I had to look at the code to understand how it was used (still not entirely sure though).

https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/63580d3e09975e90fbeabe971e3b93c44ffb2e4d/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/sac.jl#L53

This is the `target_entropy`, so if `action_dims` is positive we will have a negative entropy which seems incorrect. Not sure about the choice of default entropy as the dimensionality, that might be reasonable but I didn't find any mention of this in the linked paper (just skimmed it so might have missed it). Could be good with more information about how to set this, or maybe a link to where you can read more.

https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/63580d3e09975e90fbeabe971e3b93c44ffb2e4d/src/ReinforcementLearningZoo/src/algorithms/policy_gradient/sac.jl#L104

**Comment by findmyway** [comment]
@pilgrimygy  Could you take a look into it?

**Comment by pilgrimygy** [comment]
@albheim 
In original [SAC paper](https://arxiv.org/abs/1812.05905) on Page 8:
[![WkqfSJ.png](https://z3.ax1x.com/2021/07/13/WkqfSJ.png)](https://imgtu.com/i/WkqfSJ)
In fact, this loss is linear for α. So I did not use the optimizer.
\bar{\mathcal{H}} is the target entropy, a constant. I initialized it according to the method of the paper (on Page 17):
[![WkLGc9.png](https://z3.ax1x.com/2021/07/13/WkLGc9.png)](https://imgtu.com/i/WkLGc9)
So, it represents the dimensionality of the action. In the next version, I will fix the mistake.
Thank you!

**Comment by albheim** [comment]
> I initialized it according to the method of the paper (on Page 17):
> [![WkLGc9.png](https://camo.githubusercontent.com/b932be4cce1d5ef0fe05e81d0cc0fe57f27b3d0bb146fe338fa7cc4ba5c22018/68747470733a2f2f7a332e617831782e636f6d2f323032312f30372f31332f576b4c4763392e706e67)](https://imgtu.com/i/WkLGc9)

Ahh, thats the table I was looking for. Never went past the references...

Though I'm still a bit confused since they suggest using a negative number here, it does not seem correct for an entropy. And if we look at (18) and take the gradient wrt alpha we get -log(pi)-H where both log(pi) and H are negative, so this will always be a positive number. This means that the update, alpha -= lambda * grad J will never settle and always increase the value of alpha which seems incorrect.

**Comment by pilgrimygy** [comment]
Maybe you can refer to this: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/sac/sac.py

At the beginning of SAC training, alpha is positive, which encourages additional exploration. As the training progresses, I think the alpha should probably be smaller to exploitation. A negative alpha means that SAC is more deterministic.

**Comment by pilgrimygy** [comment]
Try reducing the learning rate. Premature deterministic strategies can lead to poor performance.

**Comment by albheim** [comment]
Looking a bit more I noticed we also have `entropy = mean(log_pi)` which seems to be the *negative* of the approximate entropy (at least if I compare to the [wiki](https://en.wikipedia.org/wiki/Entropy_(information_theory)) on information theory entropy), so having the negative of the target entropy would then make sense when comparing them. I find it confusing to call it entropy but I guess that might already be established notation.

Will try again with some other alphas and see if it works better.

**Comment by findmyway** [comment]
> I find it confusing to call it entropy 

I used to have spent a whole day debugging a related issue. (Wrongly used the sign 😭 )

**Comment by albheim** [comment]
> Though I'm still a bit confused since they suggest using a negative number here, it does not seem correct for an entropy. And if we look at (18) and take the gradient wrt alpha we get -log(pi)-H where both log(pi) and H are negative, so this will always be a positive number. This means that the update, alpha -= lambda * grad J will never settle which has to be a problem.

I also realised why my reasoning here was wrong, `log(pi)` will not always be negative as I assumed. When the variance is small for the action distribution it can become positive. The problem is that we use `mean(log(pi))` as an approximation of the (negative) entropy, which should always be negative then. This will not hold since we do small enough batches for it to sometimes stay positive, and makes the entropy plots confusing since it jumps between positive and negative entropy.

But with that said I think the implementation is correct as it is now, and the thread can be closed if you want to.
